### ISSUE TITLE
`after` sometimes not applied

### ISSUE THREAD

**Issue by Iron-E** [wbthomason/packer.nvim#435]
- `nvim --version`: NVIM v0.5.0-dev+1453-g3552916cb
- `git --version`: 2.32.0
- Operating system/version: 5.12.13-arch1-2
- Terminal name/version: `lxterminal` 0.4.0-1

### Steps to reproduce

1. Save the following file:

```lua
vim.cmd 'packadd packer.nvim'

local packer = require 'packer'

return packer.startup{function(use)
	use 'plasticboy/vim-markdown'
	use {'Iron-E/nvim-typora', after='vim-markdown'}
end}

```

2. Run `vim -u NONE <name of file> -c 'luaf %' to run the configuration
3. Run `:PackerCompile`
4. Browse the generated `packer_compiled.vim` (I'm noting the important part here):

```lua
_G.packer_plugins = {
  ["nvim-typora"] = {
    loaded = false,
    needs_bufread = true,
    path = "/home/adamz//.local/share/nvim/site/pack/packer/opt/nvim-typora"
  },
  ["vim-markdown"] = {
    -- `after = {"nvim-typora"},` is missing!
    loaded = true,
    path = "/home/adamz//.local/share/nvim/site/pack/packer/start/vim-markdown"
  }
}
```

Note that this does not happen with every plugin in my spec, but this minimal reproduction does work consistently for me.

### Actual behaviour

Plugins loading `after` were not actually loaded. This is evidenced by trying to view documentation on `nvim-typora`, or run the `:Typora` command. 

### Expected behaviour

The `after` plugins are loaded after `vim-markdown`.

**Comment by akinsho** [comment]
@Iron-E I think this is similar to #272, although tbh given that start plugins are loaded by nvim rather than something packer has control over I wonder if this can/has ever worked. I did start a PR to fix #272 over in #360, but I think this specific case can't really be fixed without making all plugins opt and having packer load them, which is discussed in another issue I forget the number of now.

**Comment by Iron-E** [comment]
I hadn't seen those issues before, thank you for linking them! I should double check if adding the `after = {…}` manually will make it work, but I sort of doubt that it will after reading those.

I'm familiar with #307— is that the approach `packer` is heading towards? I must admit, until last night I was putting all of my plugins as `opt` as well, and then using `packer.loader` to load them when I needed (mostly because of the shortcomings we've previously discussed of `cond`). This approach worked well, but it ended up making `nvim-compe` misbehave so I ditched it in favor of letting `packer` handle all the loading.

**Comment by wbthomason** [comment]
There are some good arguments for #307 (also discussed in #402, confusingly), so I'm leaning toward that approach. However, I need to figure out how to not break some setups like @clason's first (the issue being that the `packer` invocation lives in a `plugin/` file, so ensuring that plugins are sourced before they're needed becomes trickier if all plugins are `opt`). I'd be curious to hear more about how `compe` misbehaved when you took this approach (though probably in the thread at #307).

For this issue: I think this is actually a compiler bug combined with native packages being half-baked. The compiler _should_ be checking whether or not a plugin is `opt` only because of sequencing, and should load any such plugins in the compiled file (I assume, from what you say about missing commands, that there is no such `load` call in your compiled file).

The other part of it is that, as @akinsho notes, native packages mean that `vim-markdown` will always be "loaded" even if it only consists of a `ftplugin`, which means that `typora` will also always be loaded and you won't get the conditional loading you want. You could add a `ft` key manually, which should do this, but is an unpleasant hack.

We could (and I'm just speculating here) do some compile-time analysis of `ftplugin` files to identify which filetypes a given plugin activates for, then use this information to control sequence loading to match those filetypes, but it's not clear that this wouldn't create more issues (e.g. for plugins with a `ftplugin` component as well as a `plugin` component).

**Comment by akinsho** [comment]
@wbthomason the situation with trying to work around packer being invoked in the plugin dir seems like a hard to impossible problem. I don't have all the context on that situation but it sounds like an edge case since that wouldn't work with something like vim plug either.

Anyway I haven't really been following that part of the issue closely. Although is this prioritising one use case over the other numerous issues that have come up because of start plugins? I think most people load that plugins in their init

**Comment by wbthomason** [comment]
No, I still prefer to make everything `opt`, despite these edge cases. I'm just thinking about how we do this (e.g. making it configurable) to cause minimal breakage for people who do use `packer` like this (if we can do so without sacrificing the benefits of #307).

**Comment by Iron-E** [comment]
> The other part of it is that, as @akinsho notes, native packages mean that `vim-markdown` will always be "loaded" even if it only consists of a `ftplugin`, which means that `typora` will also always be loaded and you won't get the conditional loading you want.

I'm actually aiming for this; I have some other plugins that don't need `ft` (whereas `vim-markdown` might, as you noticed) but are affected by this. For example, my `nvim-tabmode` plugin must load `after` `nvim-libmodal` to work correctly. However, the `after` is not being recognized like the above.

> there is no such `load` call

Correct. I wasn't sure if one was a symptom of the other (a missing `after` or a missing `load`). But there's no instruction to load `nvim-typora` from the above case

**Comment by akinsho** [comment]
> event = 'VimEnter' in the spec of the plugin you want to load before (in this case, vim-markdown) is an excellent workaround!

This is essentially what the all opt proposal will do 😄

**Comment by oblitum** [comment]
I got another situation where `after` isn't working as I expect:

```lua
local install_path = vim.fn.stdpath('data')..'/site/pack/packer/opt/packer.nvim'
if vim.fn.glob(install_path) == '' then
    vim.fn.system { 'git', 'clone', 'https://github.com/wbthomason/packer.nvim', install_path }
end
vim.cmd [[packadd packer.nvim]]
require('packer').startup {{
    { 'wbthomason/packer.nvim', opt = true },
    { 'neoclide/coc.nvim', branch = 'release' },
    {
        'rose-pine/neovim', as = 'rose-pine',
        setup = [[vim.g.rose_pine_variant = 'moon']],
        config = [[require('rose-pine').set()]]
    },
    {
        'shadmansaleh/lualine.nvim',
        after = { 'coc.nvim', 'rose-pine' },
        config = function()
            print(vim.g.did_coc_loaded) -- THIS IS nil ON START, BUT SHOULD BE 1
        end
    }
}}
```

Starting clean, after `:PackerSync`, exiting nvim and then restarting, I get `nil` being printed on start, but it should be `1`, as [it's set to that on plugin/coc.vim](https://github.com/neoclide/coc.nvim/blob/6a9a0ee38d2d28fc978db89237cdceb40aea6de3/plugin/coc.vim#L41).

Meaning, the `config` function of lualine.nvim, which is to be loaded _after_ coc.nvim, is running before coc.nvim's `plugin/` stuff.

After `nil` gets printed on start, calling `:PackerCompile` will cause the `config` function to run again, but this time `1` gets printed, as it should be.

**Comment by oblitum** [comment]
For what is worth, changing `after` for `wants` doesn't help.
